src/main/java/org/openrefine/extensions/commons/importer/RelatedCategoryFetcher.java

```diff
@@ -0,0 +1,125 @@
+ * the file records with the list of categories they belong to
+ *
+ * @param apiUrl
+ * @param iteratorFileRecords
+ */
+public class RelatedCategoryFetcher implements Iterator<FileRecord> {
+    public static int API_TITLES_LIMIT = 20;
+    Iterator<FileRecord> iteratorFileRecords;
+    List <FileRecord> fileRecordOriginal = new ArrayList<>();
+    List<FileRecord> fileRecordNew = new ArrayList<>();
+    List<List<String>> toCategoriesColumn;
+    String apiUrl;
+    int fileRecordNewIndex = 0;
+
+    public RelatedCategoryFetcher(String apiUrl, Iterator<FileRecord> iteratorFileRecords) {
+        this.apiUrl = apiUrl;
+        this.iteratorFileRecords = iteratorFileRecords;
+    }
+
+    /*
+     * API call for fetching the related categories in batches of up to 20 titles
+     * @param list of file records
+     * @return list of related categories listed per file
+     */
+    public List<List<String>> getRelatedCategories(List <FileRecord> fileRecordOriginal) throws IOException {
+
+        String titles = fileRecordOriginal.get(0).fileName;
+        int titlesIndex =1;
+        if (titlesIndex < fileRecordOriginal.size()) {
+            titles += "|" + fileRecordOriginal.get(titlesIndex++).fileName;
+        }
+        OkHttpClient client = new OkHttpClient.Builder().build();
+        HttpUrl urlRelatedCategories = HttpUrl.parse(apiUrl).newBuilder()
+                .addQueryParameter("action", "query")
+                .addQueryParameter("prop", "categories")
+                .addQueryParameter("titles", titles)
+                .addQueryParameter("format", "json").build();
+        Request request = new Request.Builder().url(urlRelatedCategories).build();
+        Response response = client.newCall(request).execute();
+        JsonNode jsonNode = new ObjectMapper().readTree(response.body().string());
+        List<JsonNode> relatedCategories = new ArrayList<>();
+        toCategoriesColumn = new ArrayList<>();
+        for (int i = 0; i < fileRecordOriginal.size(); i++) {
+            relatedCategories.add(jsonNode.path("query").path("pages").path(fileRecordOriginal.get(i).pageId).path("categories"));
+            List<String> categoriesPerFile = new ArrayList<>();
+            for (int j = 0; j < relatedCategories.get(i).size(); j++) {
+                categoriesPerFile.add(relatedCategories.get(i).get(j).findValue("title").asText());
+            }
+            toCategoriesColumn.add(categoriesPerFile);
+        }
+
+        return toCategoriesColumn;
+    }
+
+    /*
+     * Returns {@code true} if the iteration has more elements for
+     * which to fetch related categories.
+     * (In other words, returns {@code true} if {@link #next} would
+     * return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    @Override
+    public boolean hasNext() {
+        return iteratorFileRecords.hasNext();
```

Comment:
This probably needs adapting now that your `next()` method is potentially fetching more than one FileRecord and buffering them internally.